### ISSUE TITLE
User data migrations (Discussion)

### ISSUE THREAD

**Issue by zmitchell** [flox/flox#1315]
Now that we have users, and those users have data, we need to think about how to manage migrations when we change where data lives or how that data is formatted.

Examples:
- By adding the environment registry we make the reverse links directory (`~/.local/share/flox/links`) obsolete.
- As part of that same change we shorten the length of the hashes used to refer to the locations of environments from 32 characters to 5 characters.
- We have changed the order and semantics of `hook`/`profile` scripts once, and will likely do it again in the future.
- Catalog service will require changes to the lockfile format.

Questions:
- When in the lifecycle of a `flox` invocation should we apply any migrations? Startup? Right before exit? Hidden inside a related command?
    - Example: When we do garbage collection, should we use that opportunity to remove the reverse links directory?
    - Example: When we access a floxmeta branch, should we rename it to use the shorter hashes?
    - Should we have an `apply_migrations` function that's run before we invoke the user's requested action?
- When `flox` encounters data in an old format and it knows about the new format (e.g. encounters a lockfile v1 but would produce a lockfile v2 for a new environment) do we produce data in the same (old) format or do we always produce data in the latest format?
    - Producing data in a matching (old) format ensures that we don't break compatibility with users that may be on older versions for whatever reason, but increases the support window and doesn't provide a motivation to move to newer versions.
    - Always producing the latest format could break some users, but provides a motivation to move to newer versions.

**Comment by eikenb** [comment]
So far flox is working great for me, so I'll take a minute to chime in... mostly just spitballing here so take this all with a grain of salt.

TLDR; use newest asap but support old long enough to fix issues

> When in the lifecycle of a flox invocation should we apply any migrations? Startup? Right before exit? Hidden inside a related command?

On first run after upgrade, notify of migration and do it. 

> When flox encounters data in an old format and it knows about the new format (e.g. encounters a lockfile v1 but would produce a lockfile v2 for a new environment) do we produce data in the same (old) format or do we always produce data in the latest format?

Always try to use newer format but do the migration transactional and notify of any issues in a format that would help with bug reports. Maintain runtime backwards compatibilty for N(?) releases to present a window for fixing issues. Maintain backwards compatibility for migrations for as long as possible, ie. always be able to migrate to newest. For out of runtime compatibility migrations, log issues that you would have returned for possible bug reports.

**Comment by zmitchell** [comment]
Closing this for now
